### PR TITLE
Supports advanced indexing of MultiNestedTensor (part 2)

### PR DESCRIPTION
**Pull request by weihua916**
Add concat dim=0 for `MultiNestedTensor`

Next PRs
- Column slicing
- Column concat

### CODE REVIEW

File: torch_frame/data/multi_nested_tensor.py [hunk 1]
```diff
@@ -350,13 +350,34 @@ def dtype(self) -> torch.dtype:
...
     # Static methods ##########################################################
     @staticmethod
-    def stack(xs: List['MultiNestedTensor'],
-              dim: int = 0) -> 'MultiNestedTensor':
-        # TODO: To be implemented.
-        if len(xs) == 1:
-            return xs[0]
+    def cat(xs: List['MultiNestedTensor'],
+            dim: int = 0) -> 'MultiNestedTensor':
+        if dim == 0 or dim == -3:
```
Comment (yiweny) [inline torch_frame/data/multi_nested_tensor.py:355]: May be do something like dim = dim + MultiNestedTensor.dim if dim < 0 else dim rather than having `dim == 0 or dim == -3:`

File: torch_frame/data/multi_nested_tensor.py [hunk 2]
```diff
@@ -350,13 +350,34 @@ def dtype(self) -> torch.dtype:
...
+                    raise RuntimeError(
+                        "num_cols must be the same across a list of input "
+                        "multi nested tensors.")
+            values = torch.cat([x.values for x in xs], dim=0)
+            accum_offset = 0
+            offset = []
+            for x in xs[:-1]:
+                offset.append(x.offset[:-1] + accum_offset)
+                accum_offset = accum_offset + x.offset[-1]
+            offset.append(xs[-1].offset + accum_offset)
```
Comment (yiweny) [inline torch_frame/data/multi_nested_tensor.py:369]: This operation is O(num_cols * num_rows) and then you then need to do `torch.cat` which is also quadratic. It would be inefficient. 
Can we do something like 
```
offset = torch.zeros(num_rows * num_cols + 1)
for i in range(1, num_rows * num_cols + 1):
    offset[i] = ...
```
So we can half the runtime?

File: torch_frame/data/multi_nested_tensor.py [hunk 3]
```diff
@@ -350,13 +350,34 @@ def dtype(self) -> torch.dtype:
 
     # Static methods ##########################################################
     @staticmethod
-    def stack(xs: List['MultiNestedTensor'],
-              dim: int = 0) -> 'MultiNestedTensor':
-        # TODO: To be implemented.
-        if len(xs) == 1:
-            return xs[0]
+    def cat(xs: List['MultiNestedTensor'],
```
Comment (yiweny) [inline torch_frame/data/multi_nested_tensor.py:353]: Should we support tuple here as well?

File: torch_frame/data/multi_nested_tensor.py (hunk 2 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:369]: torch.cat is not quadratic. I think this is ok.

File: torch_frame/data/multi_nested_tensor.py (hunk 1 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:355]: That'll require two lines. I think current one is simple enough.

File: torch_frame/data/multi_nested_tensor.py (hunk 3 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py]: Yes, good point. Added and tested.

File: torch_frame/data/multi_nested_tensor.py (hunk 1 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py]: Addressed.

File: torch_frame/data/multi_nested_tensor.py (hunk 2 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:371]: Hmm, I think this usage of torch.cat is pretty standard to me.

**Review by yiweny — APPROVED** [review]

File: torch_frame/data/multi_nested_tensor.py [hunk 4]
```diff
@@ -350,13 +350,36 @@ def dtype(self) -> torch.dtype:
...
-        if len(xs) == 1:
-            return xs[0]
+    def cat(xs: Union[Tuple['MultiNestedTensor', ...],
+                      List['MultiNestedTensor']],
+            dim: int = 0) -> 'MultiNestedTensor':
+        assert isinstance(xs[0], MultiNestedTensor)
+        if dim == 0 or dim + xs[0].ndim == 0:
+            num_rows = sum(x.num_rows for x in xs)
+            num_cols = xs[0].num_cols
+            for x in xs:
```
Comment (zechengz) [inline torch_frame/data/multi_nested_tensor.py:360]: nit
```suggestion
            for x in xs[1:]:
```

**Review by zechengz — APPROVED** [review]

File: torch_frame/data/multi_nested_tensor.py (hunk 4 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:360]: thanks.

File: torch_frame/data/multi_nested_tensor.py (hunk 2 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py]: Ok, I managed to make `cat` more memory efficient: https://github.com/pyg-team/pytorch-frame/pull/161/commits/aff3240c9b3416d26edb4a445d868f37c83a3bbc
